README.md
trueskill: An implementation of the Trueskill algorithm in R
============================================================

Package: trueskill <br>
Title: An implementation of the TrueSkill algorithm in R <br>

Description:  

    An R implementation of the TrueSkill Algorithm (Herbrich, R., Minka, T. and Grapel, T), 
    a Bayesian skill rating system with inference by approximate message passing on a factor graph. 
    Used by Xbox to rank gamers and identify appropriate matches.
  
    http://research.microsoft.com/en-us/projects/trueskill/default.aspx 
    
    Requires R version 3.0 as it is written with Reference Classes.
  
    URL: https://github.com/bhoung/trueskill
    
    Acknowledgements to Doug Zongker and Heungsub Lee for their python implementations 
    of the algorithm and for the liberal reuse of Doug's code comments (@dougz and @sublee on github).

    Html documentation is at http://www.bhoung.com/trueskill/
    
    Example:
    
    Team1 <- Team("T1", rank = 1, list(Player("Alice", Gaussian(25, 8.3))))
    Team2 <- Team("T1", rank = 2, list(Player("Bob", Gaussian(25, 8.3))))
    teams <- list(Team1, Team2)
    
    epsilon <- DrawMargin(draw_probability = 0.1, beta = 25 / 6, num_teams = 2)
    parameters <- Parameters(beta = 25/6, epsilon, 25 / 300)
    
    teams <- Trueskill(teams, parameters)
    
Version: 0.2 <br>
Author: Brendan Houng <brendan.houng@gmail.com> <br>
License: GPL-3 <br>
Depends: R (>= 3.0)


NEWS
trueskill 0.2
=============
* Update to support multiple players per team  
  - introduced a team class with list of players and rank;
  - removed rank from Player object, moved to Team object
  
* Using S3 method dispatch to run Trueskill on either a list of teams or a dataframe
   AdjustPlayer and Trueskill has been replaced by S3 Method Trueskill, which will
   take inputs list of teams or data.frame
   
* Added test cases also with RUnit

trueskill 0.1
=============
* trueskill algorithm to support 1 player per team 
  AdjustPlayers for once-off and Trueskill for applying trueskill to tournament data (dataframe) 



tests/1.r
#Test 1

# default inputs are
# INITIAL_MU = 25.0
# INITIAL_SIGMA = INITIAL_MU / 3.0
# INITIAL_BETA = INITIAL_SIGMA / 2.0                   
# INITIAL_GAMMA = INITIAL_SIGMA / 100.0
# DRAW_PROBABILITY = 0.10
# INITIAL_EPSILON = DrawMargin(DRAW_PROBABILITY, BETA)
  
test.examples <- function() {
  Alice  <- Player(name = "Alice",  skill = Gaussian(mu = 25, sigma = 25 / 3))
  Bob    <- Player(name = "Bob",    skill = Gaussian(mu = 25, sigma = 25 / 3))
  Chris  <- Player(name = "Chris",  skill = Gaussian(mu = 25, sigma = 25 / 3))
  Darren <- Player(name = "Darren", skill = Gaussian(mu = 25, sigma = 25 / 3))
  
  Team1 <- Team(name = "1", rank = 1, list(Alice))
  Team2 <- Team(name = "2", rank = 2, list(Bob, Chris))
  Team3 <- Team(name = "3", rank = 3, list(Darren))
  teams <- list(Team1, Team2, Team3)
  PrintPlayers(teams)
  
  epsilon <- DrawMargin(draw_probability = 0.1, beta = 25 / 6, num_teams = 3)
  parameters <- Parameters(beta = 25/6, epsilon, 25 / 300)
  
  teams <- Trueskill(teams, parameters)
  
  PrintList(teams)
  
  players <- GetPlayers(teams)
        
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2, 35.877, round(players[[1]]$skill$mu(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2,  6.791, round(players[[1]]$skill$sigma(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2, 17.867, round(players[[2]]$skill$mu(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2,  7.059, round(players[[2]]$skill$sigma(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2, 17.867, round(players[[3]]$skill$mu(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2,  7.059, round(players[[3]]$skill$sigma(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2, 21.256, round(players[[4]]$skill$mu(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2,  7.155, round(players[[4]]$skill$sigma(), 3)) 
  
}

test.deactivation <- function() {
  DEACTIVATED('Deactivating this test function')
}


tests/2.r
# Test 2.

# default inputs are 
# INITIAL_MU = 25.0
# INITIAL_SIGMA = INITIAL_MU / 3.0
# INITIAL_BETA = INITIAL_SIGMA / 2.0                   
# INITIAL_GAMMA = INITIAL_SIGMA / 100.0
# DRAW_PROBABILITY = 0.10
# INITIAL_EPSILON = DrawMargin(DRAW_PROBABILITY, BETA)

test.examples <- function() {
  Alice  <- Player(name = "Alice",  skill = Gaussian(mu = 25, sigma = 25 / 3))
  Bob    <- Player(name = "Bob",    skill = Gaussian(mu = 25, sigma = 25 / 3))
  Chris  <- Player(name = "Chris",  skill = Gaussian(mu = 25, sigma = 25 / 3))
  Darren <- Player(name = "Darren", skill = Gaussian(mu = 25, sigma = 25 / 3))

  Team1 <- Team(name = "1", rank = 1, list(Alice))
  Team2 <- Team(name = "2", rank = 2, list(Bob))
  Team3 <- Team(name = "3", rank = 2, list(Darren))
  Team4 <- Team(name = "4", rank = 3, list(Chris))
  
  teams <- list(Team1, Team2, Team3, Team4)
  
  PrintPlayers(teams)
  
  epsilon <- DrawMargin(draw_probability = 0.1, beta = 25 / 6, num_teams = 4)
  parameters <- Parameters(beta = 25/6, epsilon, 25 / 300)
  
  teams <- Trueskill(teams, parameters)
  players <- GetPlayers(teams)
  
  print(players)
  
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2, 31.716, round(players[[1]]$skill$mu(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2,  6.386, round(players[[1]]$skill$sigma(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2, 24.986, round(players[[2]]$skill$mu(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2,  5.556, round(players[[2]]$skill$sigma(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2, 25.014, round(players[[3]]$skill$mu(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2,  5.556, round(players[[3]]$skill$sigma(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2, 18.284, round(players[[4]]$skill$mu(), 3))
  checkEqualsNumeric(tolerance = .Machine$double.eps^0.2,  6.386, round(players[[4]]$skill$sigma(), 3))
  
  PrintPlayers(teams)
}

test.deactivation <- function() {
  DEACTIVATED('Deactivating this test function')
}


R/team.r
Team <- setRefClass("Team",
  fields = list(rank = "numeric", players = "list", name = "character"),
  methods = list(
    initialize = function(name = "team", rank = 1, players = list(Player(name = "player", skill = Gaussian(25, 25 / 3)))) {
      .self$players <- players
      .self$rank <- rank
      .self$name <- name
    },
    # Map of list and print?
    show = function() {
      print(sprintf("[team, rank]: [%s, %s]",  name, rank))
      print(players)   
    }    
  )
)

GetPlayers <- function(teams) {
  GetPlayer <- function(team) { return(team$players) }
  return(unlist(Map(GetPlayer, teams))) 
}

PrintPlayers = function(list) {
  for(i in 1:length(list)) {
    print(list[[i]]$players)
  }
}

PrintList = function(list) {
  for(i in 1:length(list)) {
    print(list[[i]])
  }
}               


R/player.r
#' Player: class to hold the rank, skill and names of players
#' @param rank rank of player in the match outcome
#' @param skill skill of player represented by Gaussian object e.g. Gaussian(mu = 25, sigma = 25/3)
#' @param name name the player for display purposes
#' @param ... Reference Class inheritance
Player <- setRefClass("Player",
  fields = list(name = "character", skill = "Gaussian"),
  methods = list(
    initialize = function(name = "player", skill = Gaussian$new()) {
      .self$name <- name
      .self$skill <- skill
    }, 
    UpdateSkill = function(mu, sigma) {
      .self$skill <- Gaussian$new(mu = mu, sigma = sigma)
    },
    show = function() {
      print(sprintf("[player, skill]: [%s, [(%s, %s), (%s, %s)]]", 
        name, round(skill$MuSigma()[1], 3), round(skill$MuSigma()[2], 3), round(skill$pi, 3), round(skill$tau, 3)))
    }    
  )
)


R/trueskill.r
#' function Trueskill is to be applied to tournament data (two player head to head).  
#' @param data a data frame with columns: Player, Opponent, margin
#' @description Data is required to be in long format with two rows for each match, one with player 1 first and one with player 2 first
#' Matches should be sorted such that the second copy of the match appears in the second half of the dataframe
#' The package currently only supports the trueskill algorithm with one player per team.

Trueskill.list = function(x, parameters = Parameters()) {
  
  teams <- x
  # dependencies on parameters:
  # prior factor, gamma
  # likelihood factor, beta
  # truncate factor, epsilon
  
  SortRank = function(teams) {
    GetRank = function(x) return(x$rank)
    sorted_teams <- teams[order(unlist(Map(GetRank, teams)))] 
    return(sorted_teams)
  }
  
  teams <- SortRank(teams)
  
  # skill, performance and team variable, as well as the difference variable
  # create var from player object, then map over list of players
  
  GenVar <- function(var, varname) {
    return(Variable(name = paste(varname, var$name, sep = "_")))
  }
  
  GenSumFactor <- function(team_var, perf_vars) {
    num_players <- length(perf_vars)
    coeff <- rep(list(1), num_players)
    return(SumFactor(sum_variable = team_var, term_variables = perf_vars, coeff = coeff, name = paste("SF", team_var$name, sep = "_")))
  }                                                                                    
  
  # Team Diff SumFactor
  GenTeamDiff <- function(diff_var, match_list) {
    match_name <- paste("SF", match_list[[1]]$name, "vs.", match_list[[2]]$name, sep = " ")
    return(SumFactor(sum_variable = diff_var, term_variables = match_list, coeff = list(1, -1), name = match_name))
  }
  
  # zip teams less last team, with teams less first team (t1, t2, t3) with (t2, t3, t4)
  GenTeamDiffList <- function(diff_vars, team_vars) {
    match_list <- mapply(list, team_vars[-length(team_vars)], team_vars[-1], SIMPLIFY = F)
    return(mapply(GenTeamDiff, diff_vars, match_list, SIMPLIFY = F))
  }
  
  GenTruncateFactor <- function(diff_var, player1, player2, epsilon) {
    
    if(player1$rank == player2$rank) { V <- Vdraw; W <- Wdraw }
    else { V <- Vwin ; W <- Wwin }
    
    return(TruncateFactor(diff_var, V, W, epsilon, name = paste("TF", player1$name, player2$name, sep = "_")))    
  }
  

  # create skill and perf vars for each player in each team
  # then skill and skill to perf factors
  GenSkillVars <- function(team) { 
    GenSkill <- function(player, varname) {
      return(Variable(value = player$skill, name = paste(varname, player$name, sep = "_")))
    }
    return(mapply(GenSkill, team$players, "skill")) 
  }
  
  GenPerfVars <- function(team) { return(mapply(GenVar, team$players, "perf")) }
  
  # Create each layer of factor nodes.  At the top we have priors
  # initialized to the player's current skill estimate.
  
  GenPriorFactor <- function(skill_var, player, gamma) {
    new_sigma <- sqrt(player$skill$sigma() ^ 2 + gamma ^ 2)
    param <- Gaussian(mu = player$skill$mu(), sigma = new_sigma)
    return(PriorFactor(variable = skill_var, param = param, name = paste("PF", player$name, sep = "_")))
  }
  
  GenLikelihoodFactor <- function(skill_var, perf_var, player, beta) {
    return(LikelihoodFactor(skill_var, perf_var, beta ^ 2, name = paste("LF", player$name, sep = "_")))
  }                                                               
  
  players <- GetPlayers(teams)
  
  skill_vars <- unlist(mapply(GenSkillVars, teams))
  team_perf_vars <- mapply(GenPerfVars, teams)
  
  # need to pass nested list of perf vars to GenSumFactor
  if(length(teams) == length(players)) { team_perf_vars <- Map(list, team_perf_vars) }
  perf_vars <- unlist(team_perf_vars)
  
  # create team vars and diff vars for each team 
  team_vars <- mapply(GenVar, teams, "team")
                                                                           
  team_names <- GetNames(teams)
  match_list <- mapply(list, team_names[-length(team_names)], team_names[-1], SIMPLIFY = F)
  
  GenDiffVar <- function(match_list, varname) {
    match_name <- paste(match_list[[1]], "vs.", match_list[[2]], sep = " ")
    return(Variable(name = paste(varname, match_name, sep = "_")))
  }
  
  diff_vars <- mapply(GenDiffVar, match_list, "diff")

  skill <- mapply(GenPriorFactor, skill_vars, players, parameters$gamma)
  skill_to_perf <- mapply(GenLikelihoodFactor, skill_vars, perf_vars, players, parameters$beta)

  perf_to_team <- mapply(GenSumFactor, team_vars, team_perf_vars)
  team_diff <- GenTeamDiffList(diff_vars, team_vars)
  
  # At the bottom we connect adjacent teams with a 'win' or 'draw'
  # factor, as determined by the rank values.
  
  trunc <- mapply(GenTruncateFactor, diff_vars, teams[-length(teams)], teams[-1], parameters$epsilon)
  
  # Start evaluating the graph by pushing messages 'down' from the
  # priors.
  Map(function(x) x$Start(), skill)
  Map(function(x) x$UpdateValue(), skill_to_perf)
  Map(function(x) x$UpdateSum(), perf_to_team)
  
  # Because the truncation factors are approximate, we iterate,
  # adjusting the team performance (t) and team difference (d)
  # variables until they converge.  In practice this seems to happen
  # very quickly, so I just do a fixed number of iterations.
  #
  # This order of evaluation is given by the numbered arrows in Figure
  # 1 of the Herbrich paper.

  if (length(teams) == 2) {
  		  
    z <- 1
    while (z <= 10) {
      team_diff[[1]]$UpdateSum()   # arrows (1) and (4)
      trunc[[1]]$Update()          # arrows (2) and (5)
      team_diff[[1]]$UpdateTerm(1) # arrow (3) 
      team_diff[[1]]$UpdateTerm(2) # arrow (6)
      z <- z + 1
    }          
  } else {
    z <- 1
    while (z <= 20) {
    		
      N <- length(team_diff)   
      M <- N - 1
      
      # up and right
      for (i in 1:M) {
      	team_diff[[i]]$UpdateSum()   # arrows (1)
      	trunc[[i]]$Update()          # arrows (2)
      	team_diff[[i]]$UpdateTerm(2) # arrows (3)
      }
      
      # up and left
      for (j in N:2) {
         team_diff[[j]]$UpdateSum()   # arrows (4)
         trunc[[j]]$Update()          # arrows (5)
         team_diff[[j]]$UpdateTerm(1) # arrows (6)
      }
      z <- z + 1
    }
    
    team_diff[[1]]$UpdateTerm(1)
    team_diff[[N]]$UpdateTerm(2)
    
  }
  
  # Now we push messages back up the graph, from the teams back to the
  # player skills.
  
  UpdateTerms <- function(sumfactor) {
    for (i in 1:length(sumfactor$terms)) { 
      sumfactor$UpdateTerm(i) 
    } 
  }
  # number of calls to UpdateTerm depends on number of players per team
  Map(UpdateTerms, perf_to_team)
  Map(function(x) x$UpdateMean(), skill_to_perf)
  
  # Finally, the players' new skills are the new values of the s
  # variables.
  UpdateSkill <- function(player, skill_var) {
    player$UpdateSkill(skill_var$value$mu(), skill_var$value$sigma())     
  }
  
  mapply(UpdateSkill, players, skill_vars, SIMPLIFY = F)
  
  #' display list of players nicer
  #' @param list a list of player objects
  
  return(teams)
}


Trueskill.data.frame <- function(x, parameters) {

  data <- x
  
  ApplyToRow <- function(row) {
    
    if(row$margin == 0) {
      rank1 = 1
      rank2 = 1
    } else {
      if(row$margin > 0) {
        rank1 = 1
        rank2 = 2
      } else {
        rank2 = 2
        rank2 = 1
      }
    }
  
    if(is.na(row$mu1) | is.na(row$sigma1) | is.na(row$mu2) | is.na(row$sigma2)) {
      row$mu1 <- 25                                
      row$sigma1 <- 25 / 3
      row$mu2 <- 25
      row$sigma2 <- 25 / 3 
    }
  
    Player1 <- Player(name = "P1", skill = Gaussian(mu = row$mu1, sigma = row$sigma1))
    Player2 <- Player(name = "P2", skill = Gaussian(mu = row$mu2, sigma = row$sigma2))
    
    Team1 <- Team(name = "Team1", rank = rank1, players = list(Player1))
    Team2 <- Team(name = "Team1", rank = rank2, players = list(Player2))
    
    teams <- Trueskill(list(Team1, Team2), parameters)
    players <- GetPlayers(teams)
    
    rm(Team1)
    rm(Team2)
    rm(Player1)
    rm(Player2)
                                          
    row$mu1 <- players[[1]]$skill$mu()
    row$sigma1 <- players[[1]]$skill$sigma()
    row$mu2 <- players[[2]]$skill$mu()
    row$sigma2 <- players[[2]]$skill$sigma()
    
    return(data.frame(row))  
  }
    
  N <- nrow(data) / 2
  for (i in (1:N)) {
     row <- ApplyToRow(data[i,])
     
     data[c("mu1", "sigma1", "mu2", "sigma2")][data$Player == row$Player & data$Round == row$Round,] <- row[c("mu1", "sigma1", "mu2", "sigma2")]
     data[c("mu1", "sigma1", "mu2", "sigma2")][data$Opponent == row$Player & data$Round == row$Round,] <- row[c("mu2", "sigma2", "mu1", "sigma1")]
     
     # index of the round we are up to
     next_round <- which(row$Round == levels(data$Round)) + 1
     
     if (next_round <= length(levels(data$Round))) {
  
       next_round_value <- levels(data$Round)[next_round]
       data[c("mu1", "sigma1", "mu2", "sigma2")][data$Player == row$Player & data$Round == next_round_value,] <- row[c("mu1", "sigma1", "mu2", "sigma2")]
       data[c("mu1", "sigma1", "mu2", "sigma2")][data$Opponent == row$Player & data$Round == next_round_value,] <- row[c("mu2", "sigma2", "mu1", "sigma1")]
       
     }   
  }
  
  data$mu1 <- round(data$mu1, 1)
  data$sigma1 <- round(data$sigma1, 1)
  data$mu2 <- round(data$mu2, 1)
  data$sigma2 <- round(data$sigma2, 1)
  
  return(data)           
} 

Trueskill <- function(x, parameters) UseMethod("Trueskill")

#setGeneric("Trueskill", function(x) standardGeneric("Trueskill.list"))
#setGeneric("nextNum", function(x, n) standardGeneric("nextNum"))
#setGeneric("nextNum", function(x) standardGeneric("nextNum"))

#SetMethod("Trueskill", signature(x = "data.frame", y = "missing"), function(x, y) Trueskill.data.frame(x, y))
#SetMethod("Trueskill", signature(x = "list", y = "missing"), function(x, y) Trueskill.list(x, y))




inst/example.r
#dropbox = "C:/Dropbox/"
dropbox = "/Users/brendanhoung/Dropbox/"

source(paste(dropbox, "trueskill-in-r/R/trueskill.r", sep = ""))
source(paste(dropbox, "trueskill-in-r/R/factorgraph.r", sep = ""))
source(paste(dropbox, "trueskill-in-r/R/player.r", sep = ""))
source(paste(dropbox, "trueskill-in-r/R/team.r", sep = ""))
source(paste(dropbox, "trueskill-in-r/R/init.r", sep = ""))


#setGeneric("Trueskill", function(x, ...) standardGeneric("Trueskill"))
#setMethod("Trueskill", signature(x = "data.frame"), Trueskill.data.frame(x, ...))
#setMethod("Trueskill", signature(x = "list"), Trueskill.list(x, ...))

#Trueskill <- function(x, p) UseMethod("Trueskill", x)
#Trueskill.list <- function(x, p) { Trueskill.list(teams = x, parameters = p) }
#Trueskill.data.frame <- function(x, p) { Trueskill.data.frame(data = x, parameters = p) }

#Trueskill <- function(data, parameters) UseMethod("Trueskill.data.frame", data , parameters)
#Trueskill <- function(teams, parameters) UseMethod("Trueskill.list", teams)


# Example 1.

Alice  <- Player(name = "Alice",  skill = Gaussian(mu = 25, sigma = 25 / 3))
Bob    <- Player(name = "Bob",    skill = Gaussian(mu = 25, sigma = 25 / 3))
Chris  <- Player(name = "Chris",  skill = Gaussian(mu = 25, sigma = 25 / 3))
Darren <- Player(name = "Darren", skill = Gaussian(mu = 25, sigma = 25 / 3))

Team1 <- Team(name = "1", rank = 1, list(Alice))
Team2 <- Team(name = "2", rank = 2, list(Bob, Chris))
Team3 <- Team(name = "3", rank = 3, list(Darren))
teams <- list(Team1, Team2, Team3)

epsilon <- DrawMargin(draw_probability = 0.1, beta = 25 / 6, num_teams = 3)
parameters <- Parameters(beta = 25/6, epsilon, 25 / 300)

new_teams <- Trueskill(teams, parameters)

PrintPlayers(teams)

# Example 2.

# Example 2 applies the Trueskill algorithm to a tennis tournament, the Australian Open. 
# As it is applied to 127 matches, it takes ~40 seconds to run.

# Data format of ausopen2012 is: Player, Opponent, Margin, Round, WRank, LRank
#data("ausopen2012")
attach(paste(dropbox, "trueskill-in-r/data/ausopen2012.rda", sep =""))
data <- data$data

# create match_id in order to reshape
data$match_id <- row.names(data)

# reshape wide to long on match_id such that we have
# 2 rows per match, 1 with Player1 as Player and 1 with 
# Player2 as Opponent and vice versa.

data <- data[c("Winner", "Loser", "Round", "WRank", "LRank")]
data <- reshape(data,
  idvar = "match_id",
  varying = list(c(1, 2), c(2, 1), c(4, 5), c(5,4)),
  v.names = c("Player", "Opponent", "WRank", "LRank"),
  new.row.names = 1:1000, 
  timevar = "t",
  direction = "long")
  
# data comes preformatted with winner in Player column
# set margin to 1 for win and -1 for loss.

data$margin[data$t == "1"] <- 1
data$margin[data$t != "1"] <- -1
data$t <- NULL

data$mu1 <- NA
data$sigma1 <- NA
data$mu2 <- NA
data$sigma2 <- NA

# For the first round, set Mu to 300 less the ATP rank
# Skill tends to be stable at the higher rankings (descending from 1), so set sigma at mu less mu / 3, 
# rather than the recommended mu / 3
                                
data[c("mu1","sigma1")] <- c(300 - data$WRank, round(300 - data$WRank - ((300 - data$WRank) / 3), 1))
data[c("mu2","sigma2")] <- c(300 - data$LRank, round(300 - data$LRank - ((300 - data$WRank) / 3), 1)) 

data[!data$Round == "1st Round",][c("mu1","sigma1")] <- c(NA, NA)
data[!data$Round == "1st Round",][c("mu2","sigma2")] <- c(NA, NA)

# Expects columns mu1, sigma1, mu2 and sigma2, will set mu and sigma to 25 and 25 / 3 if NA.

parameters <- Parameters()
print(parameters)

data <- Trueskill(data, parameters)
 top4 <- subset(data, Player == "Djokovic N." | Player == "Nadal R." | Player == "Federer R." | Player == "Murray A." )
 top4 <- top4[order(top4$Player,top4$Round),]

 subset(top4, Player == "Djokovic N.")      

# For a visualisation, load up our favourite package ggplot2...

library(ggplot2)
g1 <- ggplot(top4, aes(x = Round, y = mu1, group = Player, colour = Player)) + geom_point(aes(colour=factor(Player))) + geom_line(aes())       
g1

# Without having adjusted the global parameters, Trueskill does not predict match outcomes well,
# as it appears that facing stiffer opposition (higher skilled players) tends to
# diminish a player's chances of progressing in the subsequent round.

# This is consistent with commentators describing players with softer draws and playing shorter matches (3 sets as opposed to 5 sets)
# as being fresher in later rounds.          

# The other feature is that the skill of the better players is weighted towards the losing player even if the
# better player wins, so we have this effect of the 4 semifinalists having their skills dropping as 
# the tournament progresses. This could be symptomatic of high starting values, which is necessary due to some of the 
# very low rankings. E.g Lleyton Hewitt with 181.




R/competition.r
#' function Trueskill is to be applied to tournament data (two player head to head).  
#' @param data a data frame with columns: Player, Opponent, margin
#' @description Data is required to be in long format with two rows for each match, one with player 1 first and one with player 2 first
#' Matches should be sorted such that the second copy of the match appears in the second half of the dataframe
#' The package currently only supports the trueskill algorithm with one player per team
CompTrueskill <- function(data, parameters) {
	
  ApplyToRow <- function(row) {
    
    if(row$margin == 0) {
      rank1 = 1
      rank2 = 1
    } else {
      if(row$margin > 0) {
        rank1 = 1
        rank2 = 2
      } else {
        rank2 = 2
        rank2 = 1
      }
    }
  
    if(is.na(row$mu1) | is.na(row$sigma1) | is.na(row$mu2) | is.na(row$sigma2)) {
      row$mu1 <- 25                                
      row$sigma1 <- 25 / 3
      row$mu2 <- 25
      row$sigma2 <- 25 / 3 
    }
  
    Player1 <- Player(name = "P1", skill = Gaussian(mu = row$mu1, sigma = row$sigma1))
    Player2 <- Player(name = "P2", skill = Gaussian(mu = row$mu2, sigma = row$sigma2))
    
    Team1 <- Team(name = "Team1", rank = rank1, players = list(Player1))
    Team2 <- Team(name = "Team1", rank = rank2, players = list(Player2))
    
    teams <- Trueskill(list(Team1, Team2), parameters)
    players <- GetPlayers(teams)
    
    rm(Team1)
    rm(Team2)
    rm(Player1)
    rm(Player2)
                                          
    row$mu1 <- players[[1]]$skill$mu()
    row$sigma1 <- players[[1]]$skill$sigma()
    row$mu2 <- players[[2]]$skill$mu()
    row$sigma2 <- players[[2]]$skill$sigma()
    
    return(data.frame(row))  
  }
    
  N <- nrow(data) / 2
  for (i in (1:N)) {
     row <- ApplyToRow(data[i,])
     
     data[c("mu1", "sigma1", "mu2", "sigma2")][data$Player == row$Player & data$Round == row$Round,] <- row[c("mu1", "sigma1", "mu2", "sigma2")]
     data[c("mu1", "sigma1", "mu2", "sigma2")][data$Opponent == row$Player & data$Round == row$Round,] <- row[c("mu2", "sigma2", "mu1", "sigma1")]
     
     # index of the round we are up to
     next_round <- which(row$Round == levels(data$Round)) + 1
     
     if (next_round <= length(levels(data$Round))) {
  
       next_round_value <- levels(data$Round)[next_round]
       data[c("mu1", "sigma1", "mu2", "sigma2")][data$Player == row$Player & data$Round == next_round_value,] <- row[c("mu1", "sigma1", "mu2", "sigma2")]
       data[c("mu1", "sigma1", "mu2", "sigma2")][data$Opponent == row$Player & data$Round == next_round_value,] <- row[c("mu2", "sigma2", "mu1", "sigma1")]
       
     }   
  }
  
  data$mu1 <- round(data$mu1, 1)
  data$sigma1 <- round(data$sigma1, 1)
  data$mu2 <- round(data$mu2, 1)
  data$sigma2 <- round(data$sigma2, 1)
  
  return(data)           
} 


R/init.r
DrawProbability <- function(epsilon = 0.7404666, beta = 25 / 6, num_teams = 2) {
  return(2 * pnorm(epsilon / sqrt(num_teams)) * beta - 1)
}

DrawMargin <- function(draw_probability = 0.10, beta = 25 / 6, num_teams = 2) {
  return(qnorm((draw_probability + 1.0) / 2) * sqrt(num_teams) * beta)
}

# default inputs are
# INITIAL_MU = 25.0
# INITIAL_SIGMA = INITIAL_MU / 3.0
# INITIAL_BETA = INITIAL_SIGMA / 2.0                   
# INITIAL_GAMMA = INITIAL_SIGMA / 100.0
# DRAW_PROBABILITY = 0.10
# INITIAL_EPSILON = DrawMargin(DRAW_PROBABILITY, BETA)

Parameters <- setRefClass('Parameters',
  fields = list(beta = "numeric", epsilon = "numeric", gamma = "numeric"),
  methods = list(
    initialize = function(beta = 25 / 6, epsilon = DrawMargin(draw_probability = 0.1, beta = 25 / 6), gamma = 25 / 300) {
      .self$beta <- beta
      .self$epsilon <- epsilon 
      .self$gamma <- gamma
      show()
    },
    show = function() {
      print(sprintf("Parameters [(beta, epsilon, gamma)]: [(%s, %s, %s)]", 
        round(beta, 3), round(epsilon, 3), round(gamma, 3)))
    } 
  )                                                                 
)
